components/ipfs/ipfs_utils.cc

```diff
@@ -460,47 +464,110 @@ std::string GetRegistryDomainFromIPNS(const GURL& url) {
+    final_url = GURL("ipfs://" + host_parts.at(0) + url.path());
+  } else if (host_parts.at(1) == "ipns" && IsValidIPNSCID(host_parts.at(0))) {
+    final_url = GURL("ipns://" + host_parts.at(0) + url.path());
+  } else if (host_parts.at(1) == "ipns") {
+    std::string decoded = DecodeSingleLabelForm(host_parts.at(0));
+    final_url = GURL("https://" + decoded + url.path());
+  }
+
+  if (!final_url.is_valid()) {
+    return absl::nullopt;
+  }
+
     GURL::Replacements replacements;
     replacements.SetQueryStr(url.query_piece());
     replacements.SetRefStr(url.ref_piece());
     return final_url.ReplaceComponents(replacements);
-  }
+}
 
+absl::optional<GURL> ExtractSourceFromGatewayPath(const GURL& url) {
   std::vector<std::string> path_parts = base::SplitStringUsingSubstr(
       url.path(), "/", base::KEEP_WHITESPACE, base::SPLIT_WANT_NONEMPTY);
 
-  if (path_parts.size() >= 2 && path_parts.at(0) == "ipfs" &&
-      IsValidCID(path_parts.at(1))) {
-    std::string final_path;
-    if (path_parts.size() >= 3) {
-      std::vector<std::string> final_path_parts(path_parts.begin() + 2,
-                                                path_parts.end());
-      final_path = "/" + base::JoinString(final_path_parts, "/");
-    }
+  if (path_parts.size() < 2) {
+    return absl::nullopt;
+  }
+  GURL final_url;
+
+  std::string final_path;
+  if (path_parts.size() >= 3) {
+    std::vector<std::string> final_path_parts(path_parts.begin() + 2,
+                                              path_parts.end());
+    final_path = "/" + base::JoinString(final_path_parts, "/");
+  }
+
+  if (path_parts.at(0) == "ipfs" && IsValidCID(path_parts.at(1))) {
+    final_url = GURL("ipfs://" + path_parts.at(1) + final_path);
+  } else if (path_parts.at(0) == "ipns" && IsValidIPNSCID(path_parts.at(1))) {
+    final_url = GURL("ipns://" + path_parts.at(1) + final_path);
+  } else if (path_parts.at(0) == "ipns") {
+    std::string decoded = DecodeSingleLabelForm(path_parts.at(1));
+    final_url = GURL("https://" + decoded + final_path);
+  }
+
+  if (!final_url.is_valid()) {
+    return absl::nullopt;
+  }
 
-    GURL final_url = GURL("ipfs://" + path_parts.at(1) + final_path);
     GURL::Replacements replacements;
     replacements.SetQueryStr(url.query_piece());
     replacements.SetRefStr(url.ref_piece());
     return final_url.ReplaceComponents(replacements);
+}
+
+// gateway.io/ipfs/bafybeiemxf5abjwjbikoz4mc3a3dla6ual3jsgpdr4cjr3oz3evfy ->
+// ipfs://bafybeiemxf5abjwjbikoz4mc3a3dla6ual3jsgpdr4cjr3oz3evfy
+// bafybeiemxf5abjwjbikoz4mc3a3dla6ual3jsgpdr4cjr3oz3evfy.ipfs.gateway.io ->
+// ipfs://bafybeiemxf5abjwjbikoz4mc3a3dla6ual3jsgpdr4cjr3oz3evfy
+// gateway.io/ipns/en.wikipedia-on-ipfs.org ->
+// https://en.wikipedia-on-ipfs.org
+// gateway.io/ipns/en-wikipedia--on--ipfs-org ->
+// https://en.wikipedia-on-ipfs.org
+// en-wikipedia--on--ipfs-org.ipns.gateway.io ->
+// https://en.wikipedia-on-ipfs.org
```

Comment:
I would suggest making this more human readable, such as below, feel free to add more human readable texts here.
```
// Subdomain based gateway URL:
// 1) CID:
// bafybeiemxf5abjwjbikoz4mc3a3dla6ual3jsgpdr4cjr3oz3evfy.ipfs.gateway.io ->
// ipfs://bafybeiemxf5abjwjbikoz4mc3a3dla6ual3jsgpdr4cjr3oz3evfy
// 2) Single label form source:
// en-wikipedia--on--ipfs-org.ipns.gateway.io ->
// https://en.wikipedia-on-ipfs.org
//
// Path based gateway URL:
// 1) CID:
// gateway.io/ipfs/bafybeiemxf5abjwjbikoz4mc3a3dla6ual3jsgpdr4cjr3oz3evfy ->
// ipfs://bafybeiemxf5abjwjbikoz4mc3a3dla6ual3jsgpdr4cjr3oz3evfy
// 2) Single label form source:
// gateway.io/ipns/en.wikipedia-on-ipfs.org ->
// https://en.wikipedia-on-ipfs.org
// gateway.io/ipns/en-wikipedia--on--ipfs-org ->
// https://en.wikipedia-on-ipfs.org
```